nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```
@@ -0,0 +1,166 @@
+ *   * Neither the name of the <ORGANIZATION> nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/keepout_filter.hpp"
+
+#include "nav2_map_server/map_mode.hpp"
+
+namespace nav2_costmap_2d
+{
+
+KeepoutFilter::KeepoutFilter()
+: semantic_info_sub_(nullptr), map_filter_sub_(nullptr), map_filter_(nullptr)
+{
+}
+
+void KeepoutFilter::loadFilter(
+  const std::string semantic_info_topic)
+{
+  // Setting new semantic info for costmap filter subscriber
+  semantic_info_sub_ = node_->create_subscription<nav2_msgs::msg::CostmapFilterSemanticInfo>(
+    semantic_info_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::semanticInfoCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::semanticInfoCallback(
+  const nav2_msgs::msg::CostmapFilterSemanticInfo::SharedPtr msg)
+{
+  // Resetting previous subscriber each time when new semantic information arrives
+  if (map_filter_sub_) {
```

I wonder if we should throw an error instead of just accepting a new one if we already have one

---

This was made for the case if developer want to update map-filter or its parameters in run-time. Do we need to refuse such functionality?

---

I'm not sure - I'm asking you if you think that's a good idea. Certainly in that case we should have some log statement. I've noticed a lack of logging in this node

---

Got it. Added logging in https://github.com/ros-planning/navigation2/commit/fcf337566456677d0a6892b8539e902dfd086994#diff-f67a16397bb5979b7cbf4729ea53a451R69